zulip_bots: Create Link Shortener Bot.

skunkmb opened the pull request:
Hello! Here is the Link Shortener Bot I created for the *Create your own URL shortener bot* Google Code-in task. I have tested it both manually and with the automatic `BotTestCase` class.

Before use and testing, the `key` property must be set to a goo.gl API key in `link_shortener.conf` and in the `mock_config_info` call in `test_link_shortener.py`.

In use, for example, you could write

 > @**Link Shortener Bot** @**Joe Smith** See https://github.com/zulip/python-zulip-api/tree/master/zulip_bots/zulip_bots/bots for a list of all Zulip bots.

and LS Bot would respond

 > https://github.com/zulip/python-zulip-api/tree/master/zulip_bots/zulip_bots/bots: https://goo.gl/NjLZZH

Please let me know if there is something I should improve or fix! Thank you!

neiljp commented on README.md:
```diff
@@ -13,6 +13,8 @@ This repository contains the source code for Zulip's PyPI packages:
 * `zulip_botserver`: [PyPI package](https://pypi.python.org/pypi/zulip-botserver)
   for Zulip's Flask bot server.
 
+The source code is written in *Python 3*.
```
While this is a minor update, and is in a separate commit, I would suggest submitting this in a separate PR.

You should be able to cherry-pick the commit into another branch, though it is just one line :)

skunkmb replied:
Hmm. I actually *did* mention this in a separate PR: https://github.com/zulip/python-zulip-api/pull/179 but I guess it got carried over to this one too. I will remove it from the PR now.

neiljp commented on zulip_bots/zulip_bots/bots/link_shortener/link_shortener.py:
```diff
@@ -0,0 +1,68 @@
+
+        response = request.json()
+
+        if 'id' in response:
```
Could you use `.get()` here to simplify the return?

skunkmb commented:
Hi @neiljp, I removed the commit about Python 3 (it is in a [separate PR](https://github.com/zulip/python-zulip-api/pull/179)) and changed the 

```python
response = request.json()

if 'id' in response:
    return response['id']

return ''
```

to just

```python
return request.json().get('id', '')
```

as you mentioned. Are there any other changes I should make?

aero31aero commented on zulip_bots/zulip_bots/bots/link_shortener/link_shortener.py:
```diff
@@ -0,0 +1,63 @@
+            params=params
+        )
+
+        return request.json().get('id', '').
```
There is a syntax error on this line. See the trailing '.'.

skunkmb replied:
OK, I fixed it in commit 17ac82d3e930700e2d26d7bdf63d407f5cb99ed8. Thanks!

aero31aero commented on zulip_bots/zulip_bots/bots/link_shortener/test_link_shortener.py:
```diff
@@ -0,0 +1,22 @@
+            self.initialize_bot()
+
+        message = 'Shorten https://www.github.com/zulip/zulip please.'
+        response = 'https://www.github.com/zulip/zulip: https://goo.gl/6uoWKb'
```
Is the bot making an http request to the API here? You should stub that request as done in the tests for giphy bot.

skunkmb replied:
No request to the API is being made on lines 14 and 15. Since goo.gl always returns the same shortened link for the same url (as long as the REST API is being used), https://www.github.com/zulip/zulip will always be converted to https://goo.gl/6uoWKb. Because of this, `test_bot` just tests if given that `message`, the `response` is `'https://www.github.com/zulip/zulip: https://goo.gl/6uoWKb'`.

I'm not sure what "stubbing" is, can you please explain it? I looked in the giphy test and did a search for "stub" but there were no results. I also looked at some articles online, and they mentioned using "mock configurations," but I'm not entirely sure how to use those either. Thank you very much!

aero31aero replied:
Stubbing is replacing an actual function with a fake function so that we can test it in isolation. See the call to `mock_http_conversation` in the giphy tests. Also, look at the `fixtures/*.json` files that show mock http conversations.

So, yes, the line 14-15 are making a request to the API but getting the same result back. This means that we can use this result to create a mock http conversation now

skunkmb commented:
Thanks, @aero31aero! I think the reason the Travis CI build fails is because no API key is configured in the `test_link_shortener.py`. On line 10, I put a placeholder for an API key (`<your API key>`), but when a request is made with that placeholder, an error is returned. How would you recommend I proceed? Should I hard-code a key?

aero31aero commented:
You shouldn't have to hardcode a key. This is why we want to mock the http conversation and test your bot's code in isolation. I left a comment above explaining this briefly. Feel free to message on the chat if you have any doubts.

skunkmb commented:
I just pushed a new commit using a mock. However, since the goo.gl API uses POST requests, I had to add a POST request feature to the `mock_http_conversation` function in `test_lib.py`, as discussed here: https://goo.gl/fuWjf9. I made a PR with those changes at #194. Until that PR is merged, though, I don't think Travis is going to say it passes the tests, because the POST request won't be detected by `mock_http_conversation`.

aero31aero commented on zulip_bots/zulip_bots/bots/link_shortener/fixtures/test.json:
```diff
@@ -0,0 +1,19 @@
+{
```
Consider renaming this to `test_succes.json`.

aero31aero commented on zulip_bots/zulip_bots/bots/link_shortener/link_shortener.py:
```diff
@@ -0,0 +1,63 @@
+        ]
+        final_response = '\n'.join(link_pairs)
+
+        if final_response == '':
```
Consider adding some text like: `No links found. Send 'help' to see usage instructions.`. Also, add a test for the case where we don't send any link.

aero31aero commented on zulip_bots/zulip_bots/bots/link_shortener/requirements.txt:
```diff
@@ -0,0 +1 @@
+request
```
This file shouldn't be needed. Aside from the fact that as of now we don't install from each bot's requirements files, the requests module is already a dependency.

You should add it to `zulip_bots/setup.py` instead, for future proofing as follows:


```python
setuptools_info = dict(
    install_requires=[
        'zulip',
        'mock>=2.0.0',
        'html2text',  # for bots/define
    ],
)
```

aero31aero commented on zulip_bots/zulip_bots/bots/link_shortener/test_link_shortener.py:
```diff
@@ -0,0 +1,20 @@
+
+from zulip_bots.test_lib import BotTestCase
+
+class TestMessageInfoBot(BotTestCase):
```
I think you forgot to rename this class.

skunkmb commented:
@aero31aero Thanks for the feedback! I made some new commits to accommodate for those changes.

showell commented:
Looks like the tests are still failing.

skunkmb commented:
@showell I just pushed a commit to fix some linter errors (b0e0042), and now the tests pass! I think Travis just needed a new commit to make it check again, now that PR #194 has been merged.

Is there anything else I should change, or is this ready to be merged? Thanks!

showell commented:
LGTM once you squash the 2nd commit.

skunkmb commented:
@showell OK, squashed. Thanks!

neiljp commented:
I was thinking of suggesting you could annotate the bot for types, but let's save that for another PR/commit once that's discussed further elsewhere (as bots aren't checked by default by mypy, as it stands)

skunkmb commented:
@neiljp OK, let me know if you change your mind. Thanks!

neiljp commented:
@skunkmb This has been discussed and may lead to other GCI tasks:
  https://chat.zulip.org/#narrow/stream/integrations/topic/Partial.20Mypy.20annotation

It is certainly not currently a requirement for bots, and I believe your bot has been provisionally included to be typed in the new GCI tasks, but if you wanted to add type annotations I don't imagine there'd be objections - though you could keep them in a separate commit to make it easier to cherry-pick into another task/PR later.

skunkmb commented:
@neiljp Thanks! Maybe after the task I'm working on now (Chess Bot), I'll create a new PR adding the type annotations.

showell commented:
Merged, thanks @skunkmb!
